[WIP] Add profile picture in message popover.

hackerkid opened the pull request:
Discussion here - https://chat.zulip.org/#narrow/stream/frontend/topic/profile.20picture
![screenshot from 2017-02-24 01 18 18](https://cloud.githubusercontent.com/assets/7190633/23276155/439a4ca0-fa2f-11e6-9acb-5f4dc572c0cf.png)

timabbott commented:
I merged the first commit after extending the test slightly as b7a6826fda1f296ae59881618de75653b7566d02

hackerkid commented:
Updated screenshot

![screenshot from 2017-03-02 22 53 02](https://cloud.githubusercontent.com/assets/7190633/23518925/17ff835a-ff9b-11e6-95cb-f9eb63ba1b01.png)

timabbott commented:
We should probably figure out CSS to fix the alignment of the icon/text pairings; ideally, the icons would all be center-aligned in one column, and the text would all be left-aligned in a second.

hackerkid commented:
Working on the CSS :)

hackerkid commented:
Updated the CSS.

![screenshot from 2017-03-04 02 37 57](https://cloud.githubusercontent.com/assets/7190633/23569106/ae874d1a-0083-11e7-9861-e0992b2e4756.png)

hackerkid commented:
@timabbott Please review when you are free.

hackerkid commented:
I replaced the user_side_bar action popup with message_info popup. Now both the popups are identical.

timabbott commented:
This is looking awesome!  I merged the first few commits.  A few problems:
* The shape of the image in the popover doesn't look good with gravatars, because gravatars are square.  I think we should just so a square space for the profile image; what do you think?  The alternative would be to try to cut off equally from the top and bottom, at least, rather than just the bottom.
* The "user info" popover (as popped from the right sidebar) doesn't close if you click e.g. "reply mentioning user"
* "Remove title of message info popup." is missing part of the commit, I think?  It doesn't change the code that actually uses that template...
* "Remove arrow from message info popover." could benefit from an explanatory comment, since it's kinda a hack :).  Might be worth putting the HTML content into a template and reading it in, just for clarity and greppability.
* For "Make message info popover position fixed.", I think we might want to have some way to make the location vary; I take it that's technically somewhat hard to do?  The "user" avatar in particular seems to be placed well or badly depending on screen size, and I bet wouldn't work well in a narrow window where the user list is hidden or on the left...

hackerkid commented:
@timabbot > the alternative would be to try to cut off equally from the top and bottom, at least, rather than just the bottom.

I made some minor change in CSS so that it's cutting off equally. I also asked the folks in Zulip which one they prefer more (rectangle or square). https://chat.zulip.org/#narrow/stream/design/topic/User.20Info.20Popover

hackerkid commented on static/styles/zulip.css:
```diff
@@ -2532,3 +2532,25 @@ ul.sender_info_popover i {
+.popover-avatar {
+    height: 220px;
+    width: 300px;
+    background-size: cover;
```
This should cut off the image equally from top and bottom.

hackerkid commented:
@timabbott 
> The shape of the image in the popover doesn't look good with gravatars, because gravatars are square. I think we should just so a square space for the profile image; what do you think? The alternative would be to try to cut off equally from the top and bottom, at least, rather than just the bottom.

I changed the avatar box to square. Also the avatar are centered now so the cutoff should be uniform from the opposite sides.

> The "user info" popover (as popped from the right sidebar) doesn't close if you click e.g. "reply mentioning user"

Fixed this.

> "Remove title of message info popup." is missing part of the commit, I think? It doesn't change the code that actually uses that template...

Actually in the next commit we are inserting the avatar in that template. That's why I didn't remove the use of the template from popover.js

> "Remove arrow from message info popover." could benefit from an explanatory comment, since it's kinda a hack :). Might be worth putting the HTML content into a template and reading it in, just for clarity and greppability.

I am using a template now instead of passing HTML. Also changed the commit message.

> For "Make message info popover position fixed.", I think we might want to have some way to make the location vary; I take it that's technically somewhat hard to do? The "user" avatar in particular seems to be placed well or badly depending on screen size, and I bet wouldn't work well in a narrow window where the user list is hidden or on the left...

I changed it so that only the height is fixed now. The Y position is back to it used to be.

hackerkid commented:
The avatar is now square. 

![screenshot from 2017-03-06 05 05 21](https://cloud.githubusercontent.com/assets/7190633/23592671/9a7a3bf4-022a-11e7-9bb3-edf4d214694e.png)

hackerkid commented:
@timabbott Ready for another review.

timabbott commented:
The vertical placement for the right sidebar-based popover looks pretty good; I think I'd just move the horizontal placement over to the left a little so that it just barely doesn't overlap the right sidebar at all, since this boundary doesn't look good:

![image](https://cloud.githubusercontent.com/assets/2746074/23690583/a0689508-0377-11e7-9e24-86d9cab2973a.png)

I feel like the one on messages isn't working that well.  Here's a proposal for that case:
* If you clicked the avatar, the x-position is always directly to the right of the avatar, and the y-position has it going down from the top of the avatar, but no lower than the bottom of the screen.  
* If you clicked the person's name, the same algorithm, but placed to the right of their name rather than the right of their avatar.

Is that doable?  I think with those changes this design would feel super awesome.

timabbott commented:
@hackerkid I noticed you pushed a new version; did that resolve any of the issues?

hackerkid commented:
@timabbott Nope. I fixed some merge conflicts.

hackerkid commented:
@timabbott I got a chance to work on this while on the train. Here is how the right side popover looks now. 

![screenshot from 2017-03-22 19 04 53](https://cloud.githubusercontent.com/assets/7190633/24200226/76f08e94-0f32-11e7-85d7-e864595dd5fb.png)

Regarding the message_info popover. I made it work in such a way that popup is always placed in the right side of the user_name if it does not overflow the screen. If it overflows, the code will try to place the popup in bottom of the username. If that is not possible then it tries to keep the popup in the top of the user_name. If that too fails it's just keep the popup in the right side of the username. Bootstrap only allows limited option for positioning the popup. I can try for some CSS hacks if this does not do the job.

hackerkid commented on static/js/popovers.js:
```diff
@@ -81,9 +81,21 @@ function show_message_info_popover(element, id) {
             private_message_class: "respond_personal_button",
         };
 
-        var ypos = elt.offset().top - message_viewport.scrollTop();
```
`message_viewport.scrollTop()` always returns values greater than `elt.offset().top` once the user scrolls down the initial messages that was loaded on the screen. So `ypos` was always negative. Hence the popup was always placed in bottom.

timabbott replied:
Makes sense!

timabbott commented:
Nice, I think this positioning is good enough to merge; we can see whether we need to do another round of CSS work based on feedback :).

A few follow-ups issues to think about:
* #1506.  @umairwaheed and @hackerkid you make sure you agree on who is doing the frontend for this (both for setting the timezone and displaying it in the profile if it is set).
* Custom profile field UI (backend is still pending, though)
* #4251.
* We should probably move the "message sent while you weren't subscribed" / historical check from this popover to "message actions" popover (or maybe a new one you get by clicking the timestamp?) for clarity.

I'd probably recommend working on the timezone one next.
